### PR TITLE
Datasource id is required if multiple datasource is enabled and no default cluster supported

### PR DESCRIPTION
**Pull request by xinruiba**
### Description

When datasource_id is empty and default cluster not supported, the search returns 500 error which is unexpected since this is more of input error. This change is to check if multiple datasource enabled and default cluster not supported, when search with a empty datasource_id then the request is invalid and should return code 400.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5754
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5699

## Testing the changes

1. Send empty datasource_id in the payload with datasource enabled and defaultCluster disabled in the yaml config returns 400 instead of 500

        Request:
        `curl 'localhost:5601/internal/search/opensearch' --data '{"params": {"index":"movies*","body":{},"preference":1697624613679}, "dataSourceId": ""}' -H 'x-amzn-aosd-app-id:random' -H 'osd-xsrf:osd-fetch' -H 'Content-Type:application/json'`

        Response:
        `{"statusCode":400,"error":"Bad Request","message":"Data Source Error: Data source id is required when no openseach hosts config provided","attributes":{"error":"Data Source Error: Data source id is required when no openseach hosts config provided"}}`

2. Send payload with no datasource_id in the payload with datasource enabled and defaultCluster disabled in the yaml config  returns 400 instead of 500

        Request:
        `curl 'localhost:5601/internal/search/opensearch' --data '{"params": {"index":"movies*","body":{},"preference":1697624613679}}' -H 'x-amzn-aosd-app-id:random' -H 'osd-xsrf:osd-fetch' -H 'Content-Type:application/json'`

        Response:
        `{"statusCode":400,"error":"Bad Request","message":"Data Source Error: Data source id is required when no openseach hosts config provided","attributes":{"error":"Data Source Error: Data source id is required when no openseach hosts config provided"}}`

3. Send payload with invalid datasource_id

        Request:
        `curl 'localhost:5601/internal/search/opensearch' --data '{"params": {"index":"movies*","body":{},"preference":1697624613679}, "dataSourceId": "invalid"}' -H 'x-amzn-aosd-app-id:random' -H 'osd-xsrf:osd-fetch' -H 'Content-Type:application/json'`

        Response:
        `{"statusCode":404,"error":"Not Found","message":"Data Source Error: Data Source Error: Saved object [data-source/invalid] not found","attributes":{"error":"Data Source Error: Data Source Error: Saved object [data-source/invalid] not found"}}`

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: config/opensearch_dashboards.yml [hunk 1]
```diff
@@ -237,6 +237,8 @@
 
 # Set the value of this setting to true to enable multiple data source feature.
 #data_source.enabled: false
+# Set the value of this setting to true to enable default cluster in data source feature.
+#data_source.defaultCluster: false
```
Comment (SuZhou-Joe) [inline config/opensearch_dashboards.yml:241]: Default should be true from the `decideClient` logic?

Comment (xinruiba) [reply config/opensearch_dashboards.yml:241]: Good catch, will update.

File: src/plugins/data/server/search/opensearch_search/decide_client.ts [hunk 2]
```diff
@@ -9,14 +9,17 @@ import { IOpenSearchSearchRequest } from '..';
...
+  const dataSourceClient =
     request.dataSourceId && context.dataSource
       ? await context.dataSource.opensearch.getClient(request.dataSourceId)
-      : withLongNumeralsSupport
-      ? context.core.opensearch.client.asCurrentUserWithLongNumeralsSupport
-      : context.core.opensearch.client.asCurrentUser;
-  return client;
+      : defaultOpenSearchClient;
+
+  return withDataSourceEnabled ? dataSourceClient : defaultOpenSearchClient;
```
Comment (BionIT) [inline src/plugins/data/server/search/opensearch_search/decide_client.ts:24]: Would it be cleaner to change line 19-24 to `return withDataSourceEnabled && request.dataSourceId && context.dataSource ? await context.dataSource.opensearch.getClient(request.dataSourceId) : defaultOpenSearchClient;`

File: src/plugins/data/server/search/opensearch_search/decide_client.ts [hunk 3]
```diff
@@ -9,14 +9,17 @@ import { IOpenSearchSearchRequest } from '..';
...
+    : context.core.opensearch.client.asCurrentUser;
+
+  const dataSourceClient =
     request.dataSourceId && context.dataSource
       ? await context.dataSource.opensearch.getClient(request.dataSourceId)
-      : withLongNumeralsSupport
-      ? context.core.opensearch.client.asCurrentUserWithLongNumeralsSupport
-      : context.core.opensearch.client.asCurrentUser;
-  return client;
+      : defaultOpenSearchClient;
```
Comment (BionIT) [inline src/plugins/data/server/search/opensearch_search/decide_client.ts:22]: If default cluster is true while data source plugin is enabled, we should use the default opensearch client, right?

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts [hunk 4]
```diff
@@ -92,7 +105,7 @@ export const opensearchSearchStrategyProvider = (
       } catch (e) {
         if (usage) usage.trackError();
 
-        if (dataSource && request.dataSourceId) {
+        if (dataSource) {
```
Comment (BionIT) [inline src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts:108]: should the check on `dataSource?.dataSourceEnabled()`

**Review by BionIT — COMMENTED** [review]
Left some comments about the feature flag

File: src/plugins/data/server/search/opensearch_search/decide_client.ts (hunk 2 above)
Comment (xinruiba) [reply src/plugins/data/server/search/opensearch_search/decide_client.ts:24]: Thanks for the comment, will update.

File: src/plugins/data/server/search/opensearch_search/decide_client.ts (hunk 3 above)
Comment (xinruiba) [reply src/plugins/data/server/search/opensearch_search/decide_client.ts:22]: It depends on dataSourceId.

If DataSourceId not exist or empty, we will use default opensearch client.
If DataSourceId exist, we will try to get MD client to sever that DataSourceId. And 404 might be returned if dataSourceId is not correct.

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts (hunk 4 above)
Comment (xinruiba) [reply src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts:108]: That's a good point. Will update. Thanks

File: src/plugins/data/server/search/opensearch_search/decide_client.ts (hunk 3 above)
Comment (BionIT) [reply src/plugins/data/server/search/opensearch_search/decide_client.ts:22]: Got it!

**Comment by bandinib-amzn** [comment]
I see this issue is introduced when config option added for hide/show local cluster option in data source picker. 
1. I would move the change of adding feature flag to PRs addressing #5754
2. Also I would like to merge this PR after #5754 as this issue only happens when  both`data_source.enabled` and 'defaultCluster` are disabled (value of both settings are false). If there is no hide/show local cluster change, then I don't see the issue as local cluster id will always gets passed.

File: config/opensearch_dashboards.yml (hunk 1 above)
Comment (xinruiba) [reply config/opensearch_dashboards.yml]: Updated, resolving comments

File: src/plugins/data/server/search/opensearch_search/decide_client.ts (hunk 2 above)
Comment (xinruiba) [reply src/plugins/data/server/search/opensearch_search/decide_client.ts]: Updated, thanks. Resolving this comment.

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts (hunk 4 above)
Comment (xinruiba) [reply src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts]: Updated, thanks.
Resolving this comment

**Comment by xinruiba** [comment]
> I see this issue is introduced when config option added for hide/show local cluster option in data source picker.
> 
> 1. I would move the change of adding feature flag to PRs addressing [Add a new feature flag under data_source to enable/disabled default cluster  #5754](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5754)
> 2. Also I would like to merge this PR after [Add a new feature flag under data_source to enable/disabled default cluster  #5754](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5754) as this issue only happens when  both`data_source.enabled` and 'defaultCluster` are disabled (value of both settings are false). If there is no hide/show local cluster change, then I don't see the issue as local cluster id will always gets passed.

Thanks for the comments, I create a quick one pager, let's review it first~

**Review by Flyingliuhub — APPROVED** [review]

**Comment by xinruiba** [comment]
> I see this issue is introduced when config option added for hide/show local cluster option in data source picker.
> 
> 1. I would move the change of adding feature flag to PRs addressing [[MD] Able to Enabled/Disable "Local Cluster" option from datasource DropDown and Index Pattern creation page #5754](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5754)
> 2. Also I would like to merge this PR after [[MD] Able to Enabled/Disable "Local Cluster" option from datasource DropDown and Index Pattern creation page #5754](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5754) as this issue only happens when  both`data_source.enabled` and 'defaultCluster` are disabled (value of both settings are false). If there is no hide/show local cluster change, then I don't see the issue as local cluster id will always gets passed.

Based on the one pager, this change is for server error handling, and able to decoupled from other UX changes. 
Synced with team offline and we think this PR should be able to merge in. Thanks

**Review by ZilongX — APPROVED** [review]

**Comment by kavilla** [comment]
Not required for 2.12

**Pull request merged**